File: 0931-minimum-falling-path-sum/0931-minimum-falling-path-sum.cpp
class Solution {
public:
    int minFallingPathSum(vector<vector<int>>& matrix) {
        vector<vector<int>>dp(matrix.size(),vector<int>(matrix[0].size(),INT_MAX));
            for(int i=0;i<matrix[0].size();i++)
            {
                dp[0][i]=matrix[0][i];
            }
            for(int i=1;i<matrix.size();i++)
            {
                for(int j=0;j<matrix[0].size();j++)
                {
                    if(j>0)
                        dp[i][j]=min(dp[i-1][j-1]+matrix[i][j],dp[i][j]);
                     if(j<matrix[0].size()-1)
                        dp[i][j]=min(dp[i-1][j+1]+matrix[i][j],dp[i][j]);
                     dp[i][j]=min(dp[i-1][j]+matrix[i][j],dp[i][j]);
                }
            }
        int ans=INT_MAX;
        for(int i=0;i<matrix[0].size();i++)
        {
            ans=min(dp[matrix.size()-1][i],ans);
        }
        return ans;
        
    }
};

File: 0931-minimum-falling-path-sum/0931-minimum-falling-path-sum.java
class Solution {
    public int minFallingPathSum(int[][] matrix) {
        int [][]dp=new int [matrix.length][matrix[0].length];
        int ans=Integer.MAX_VALUE;
        for(int i=0;i<matrix.length;i++){
            for(int j=0;j<matrix[0].length;j++){
                if(i==0){
                    dp[i][j]=matrix[i][j];
                }else{
                    int left=j>0?dp[i-1][j-1]:Integer.MAX_VALUE;
                    int right=j<matrix[0].length-1?dp[i-1][j+1]:Integer.MAX_VALUE;
                    int up=dp[i-1][j];
                    dp[i][j]=matrix[i][j]+Math.min(Math.min(left,right),up);

                }
                if(i==matrix.length-1){
                        ans=Math.min(ans,dp[i][j]);
                }
                // System.out.print(dp[i][j]+" ");
            }
            // System.out.println();
        }
        return ans;
        
    }
}